### PR TITLE
HPCC-22944 Roxie incorrectly copying files to hpcc-data2 in "simple" mode

### PR DESCRIPTION
**Pull request by richardkchapman**
Incorrectly calculating replication level in "simple" replication mode.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Compiled.

Documentation that needs changing (regarding how to specify replication level when launching  new instances in a dynamic topology setting) has not been written yet afaik.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@mckellyln Please check

**Comment by mckellyln** [comment]
Tested ok, will also look at code

File: roxie/ccd/ccdmain.cpp [hunk 1]
```diff
@@ -647,7 +647,15 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
...
                 {
-                    slaveChannels.push_back(atoi(value.c_str()));
+                    char *tail = nullptr;
+                    unsigned channel = strtoul(value.c_str(), &tail, 10);
+                    unsigned repl = 0;
+                    if (*tail==':')
+                    {
+                        tail++;
+                        repl = atoi(tail);
+                    }
```
Comment (mckellyln) [inline roxie/ccd/ccdmain.cpp]: If tail wasn't ':' then I suppose 0 is the right replication value ?  Seems so.

File: roxie/udplib/udptopo.cpp [hunk 2]
```diff
@@ -119,18 +120,26 @@ CTopologyServer::CTopologyServer(const char *topologyInfo)
     {
         StringArray fields;
         fields.appendList(line.c_str(), "|", true);
-        if (fields.length()==3)
+        if (fields.length()==4)
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/udplib/udptopo.cpp]: All fields are mandatory in this (self-defined, internal) message format. There's no reason to complicate the logic by making that not so

**Comment by mckellyln** [comment]
@richardkchapman looks good

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

File: roxie/ccd/ccdmain.cpp [hunk 3]
```diff
@@ -1128,13 +1136,13 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
...
+                RoxieEndpointInfo me = {RoxieEndpointInfo::RoxieServer, 0, { (unsigned short) port, myIP }, 0};
                 myRoles.push_back(me);
             }
-            for (unsigned channel: slaveChannels)
+            for (std::pair<unsigned, unsigned> channel: slaveChannels)
             {
                 mySlaveEP.set(ccdMulticastPort, myIP);
-                RoxieEndpointInfo me = { RoxieEndpointInfo::RoxieSlave, channel, mySlaveEP };
+                RoxieEndpointInfo me = { RoxieEndpointInfo::RoxieSlave, channel.first, mySlaveEP, channel.second };
                 myRoles.push_back(me);
```
Comment (ghalliday) [inline roxie/ccd/ccdmain.cpp]: minor: would this be better using emplace_back?

Comment (richardkchapman) [reply roxie/ccd/ccdmain.cpp]: I think you suggested that the first time I wrote this code, and I could not come up with any way to get it to compile.

It's code where clarity is far more important than efficiency (executed once at startup) anyway.

Comment (ghalliday) [reply roxie/ccd/ccdmain.cpp]: ` myRoles.emplace_back(RoxieEndpointInfo{ RoxieEndpointInfo::RoxieSlave, channel.first, mySlaveEP, channel.second });
`

seems to work, but as you say efficiency isn't significant in this piece of code.

**Pull request merged**